docs/source/spec-workload-executor.rst

```diff
@@ -37,55 +37,137 @@ Behavioral Description
+   
+   * failure lists: arrays of documents describing the failures that occurred
+     while the workload executor was executing the operations. Each client
+     entity may report errors to a separate failure list, or the same
+     failure list may be used by multiple client entities.
+   
+   * event lists: arrays of documents describing the events that occurred
+     while the workload executor was executing the operations. Each client
+     entity may report events to a separate event list, or the same
+     event list may be used by multiple client entities.
+
+#. MUST calculate the aggregate counts of errors (``numErrors``) and failures
+   (``numFailures``) from the error and failure lists.
+
+#. MUST write the collected events, errors and failures into a JSON file named
+   ``events.json`` in the current directory
+   (i.e. the directory from where the workload executor is being executed). 
+   The data written MUST be a map with the following fields:
+   
+   - For each event list entity, the name of the entity MUST become a key and the
+     documents stored in the entity MUST become the respective value.
+   
+   - For each error list entity, the name of the entity MUST become a key and the
+     documents stored in the entity MUST become the respective value.
+   
+   - The errors that the workload executor handles MUST be stored using the
+     ``errors`` key.
+   
+   - For each failure list entity, the name of the entity MUST become a key and the
+     documents stored in the entity MUST become the respective value.
+   
+   - The failures that the workload executor handles MUST be stored using the
+     ``failures`` key.
+         
+#. MUST write the collected workload statistics into a JSON file named
+   ``results.json`` in the current working directory (i.e. the directory
+   from where the workload executor is being executed). Workload statistics
+   MUST contain the following fields (drivers MAY report additional statistics
+   using field names of their choice):
+
+   * ``numErrors``: the number of operation errors that were encountered
+     during the test. This includes errors handled by the workload executor
+     and errors handled by the unified test runner.
+   * ``numFailures``: the number of operation failures that were encountered
+     during the test. This includes failures handled by the workload executor
+     and failures handled by the unified test runner.
+   * ``numSuccesses``: the number of successful loop iterations executed
+     during the test.
+
+ .. note:: The values of ``numErrors`` and ``numFailures`` are used by ``astrolabe`` to determine the overall
+    success or failure of a driver workload execution. A non-zero value for either of these fields is construed
+    as a sign that something went wrong while executing the workload and the test is marked as a failure.
+    The workload executor's exit code is **not** used for determining success/failure and is ignored.
```

Comment:
Treating both failures and errors the same here (i.e. "test is marked as a failure") may support my earlier question about why we should differentiate them at all.